### ISSUE TITLE
Function Factory Providers (aka Binding) Issue 

### ISSUE THREAD

**Issue by ritch** [strongloop/loopback-next#269]
@bajtos @superkhau 

Consider this `Provider` (aka Binding)

```ts
export class AuthenticatedUser {
  constructor(
    private @inject('authentication.required') required,
    private @inject('http.request') req,
    private @inject('authentication.strategy') strategy,
  )
  static name = 'authentication.user';
  async value() {
    const user = await this.strategy.authenticate(this.req);

    // ...

    return user;
  }
}
```

...and this `Sequence`

```ts
class TodoSequence extends Sequence {
  constructor(
    public @injectAuthenticate() authenticate,
    public @injectAuthorize() authorize,
    public @injectInvoke() invoke,
    public @injectSend() send,
    public @injectReject() reject
  ) {}

  async run() {
    if(false) {
      // this will never execute
      await this.authenticate();
    }
    await this.authorize();
    this.send(await this.invoke());
  }
}
```

> await this.authenticate();

This code will never run, but the authentication.user value function will. We need some way to ensure that only when `authenticate()` is called that we resolve the `authentication.user` binding.

```ts
export class Authenticate {
  constructor(
    private @inject('authentication.required') required,
    // this will will be resolved - even if the `authenitcate()` function is never called
    private @inject('authentication.user') user
  )
  static name = 'authentication.user';
  value() {
    console.log(this.user); // {username: ...};
    return () => {
        const isAuthRequired = this.required;
        const user = this.user;

        if (isAuthRequired && !user) {
          throw ...;
        }
    }
  }
}
```

**Comment by superkhau** [comment]
I'm on the fence regarding keeping the simplicity of always returning values in the providers versus simply using a function (so `authenticate.required` doesn't get resolved immediately).

IMO, we should just return a function for now (to make it work) and see how it plays out after a few more component implementations to see if this scenario is more common than we think. We can make a decision on after getting some third-party devs to try it out.

**Comment by bajtos** [comment]
> This code will never run, but the authentication.user value function will. We need some way to ensure that only when authenticate() is called that we resolve the authentication.user binding.

Maybe `authentication.user` should behave like a property that's initially set to `undefined` or `null`, and the code implementing `this.authenticate` should change (rebind?) the value to the actual user.

My proposal may be easier to implement if `authentication` binding behaved like a real object.

```ts
interface Authentication<User>: {
  required: boolean;
  user?: User;
}

// a mock-up, maybe we should use a proper class instance
const authentication: Authentication = {
  required: false,
  user: undefined
};

ctx.bind('authentication').to(authentication);
// mockup implementation that does not handle sync/async
// ideally, property getters should be provided by our IoC framework
ctx.bind('authentication.required').toDynamicValue(ctx => ctx.get('authentication').required);
ctx.bind('authentication.user').toDynamicValue(ctx => ctx.get('authentication').user);

// `this.authenticate`
ctx.bind('authenticate').toDynamicValue(ctx => 
  // do we support argument injection for dynamic-value getters yet?
  authenticate(ctx.get('authentication'),...)
);

function authenticate(auth: Authentication) {
  auth.user = await this.strategy.authenticate(this.req);
}
```

The downside of the above code is that `authentication.user` bindings resolved before `this.authenticate()` is called will see the default `undefined/` user. It seems to me that we either need to 
 1. start defining phases to allow us to update some bindings (like `authentication.user`) before they are injected into controller methods (and possible other places like `this.authorize`) or
 2. change the design of our binding to bind/inject the full `authorization` object, which is the only way in JavaScript to passing a reference to things, as least AFAIK.

I think the first option is violating our intent to avoid any static phases, so I guess we need to go with the second one. Unless we can find another approach?

**Comment by bajtos** [comment]
Let me rephrase my previous comment.

The idea is to treat `authentication.user` as a (mutable) reference type:
 - In the initial Dependency Injection phase, all classes and sequence steps get injected the reference.
 - `this.authenticate()` step changes the value the reference points to, from `undefined` to a user instance
 - Then the controller method is invoked. When it accesses `authentication.user`, it resolves the reference to the actual value. The important part is that the final value is resolved at the time the method is accessing the current user, not at the time when dependencies are injected.

In JS/TS, we can pass mutable references as object properties:

```js
function MyControllerMethod(auth) {
  return () => console.log('Current user:', auth.user);
}

function AuthStrategy(auth) {
  return () => auth.user = 'admin';
}

const auth = {};
// inject dependencies 
// (manually, to keep this example simpler)
const authenticate = AuthStrategy(auth);
const invoke = MyControllerMethod(auth);

// run the sequence
authenticate();
invoke();

// output:
// Current user: admin
```

**Comment by raymondfeng** [comment]
I think activities (such as `authenticate`) in the request processing sequence should have the ability to bind new information into the context so that downstream steps can access them via the container.

**Comment by bajtos** [comment]
@ritch @raymondfeng @superkhau @deepakrkris is this still relevant?

**Comment by bajtos** [comment]
This has been addressed by adding `@inject.getter` and `@inject.setter` decorators, see #492 and #494
